BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c

```diff
@@ -380,6 +382,117 @@ UpdateMadt (
   return EFI_SUCCESS;
 }
 
+/**
+  This function updates FWST ACPI data structure based on ESRT table.
+
+  @param[in]  Current               Pointer to MADT table
+  @retval     EFI_SUCCESS           Patched the MADT
+              EFI_UNSUPPORTED       Sysinfo table not found.
+              EFI_OUT_OF_RESOURCES  If less MADT entries
+**/
+EFI_STATUS
+UpdateFwst (
+  IN UINT8   *Current
+  )
+{
+  UINT32        RsvdBase;
+  UINT32        EsrtTableSize;
+  UINT64        OemId;
+  EFI_STATUS    Status;
+  LOADER_GLOBAL_DATA              *LdrGlobal;
+  BOOT_LOADER_VERSION             *Version;
+  EFI_SYSTEM_RESOURCE_TABLE       *pEsrtTable;
+  EFI_SYSTEM_RESOURCE_ENTRY       *pEsrtEntry;
+  FIRMWARE_UPDATE_STATUS          *FwUpdateStatus;
+  EFI_FWST_ACPI_DESCRIPTION_TABLE *pFwstAcpiTable;
+
+  LdrGlobal = (LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer();
+
+  //
+  // We do not need to populate ACPI table during firmware update
+  //
+  if (LdrGlobal->BootMode == BOOT_ON_FLASH_UPDATE) {
+    return EFI_SUCCESS;
+  }
+
+  //
+  // Get bootloader reserved region base
+  //
+  Status = GetComponentInfo (FLASH_MAP_SIG_BLRESERVED, &RsvdBase, NULL);
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  //
+  // Allocate memory for ESRT structure
+  //
+  EsrtTableSize = sizeof (EFI_SYSTEM_RESOURCE_TABLE_HEADER) + sizeof (EFI_SYSTEM_RESOURCE_ENTRY);
+  pEsrtTable = AllocatePages (EFI_SIZE_TO_PAGES (EsrtTableSize));
+  if (pEsrtTable == NULL) {
+    DEBUG((DEBUG_ERROR, "Could not allocate resources for esrt table\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  ZeroMem (pEsrtTable, EsrtTableSize);
+
+  //
+  // Fill in ESRT table structure
+  //
+
+  //
+  // Fill in ESRT header
+  // Currently only system firmware is supported
+  //
+  pEsrtTable->Header.FwResourceCount = 0x1;
+  pEsrtTable->Header.FwResourceCountMax = 0x1;
+  pEsrtTable->Header.FwResourceVersion = EFI_SYSTEM_RESOURCE_TABLE_FIRMWARE_RESOURCE_VERSION;
+
+  //
+  // Fill in ESRT entry for system firmware
+  //
+  pEsrtEntry = (EFI_SYSTEM_RESOURCE_ENTRY *)((UINTN)pEsrtTable + sizeof(EFI_SYSTEM_RESOURCE_TABLE_HEADER));
+
+  CopyMem(&pEsrtEntry->FwClass, &gEsrtSystemFirmwareGuid, sizeof (EFI_GUID)); 
+  pEsrtEntry->FwType = ESRT_FW_TYPE_SYSTEMFIRMWARE;
+
+  //
+  // Get current version and lowest supported SVN version
+  //
+  if (LdrGlobal->VerInfoPtr != NULL) {
+    Version = LdrGlobal->VerInfoPtr;
+    pEsrtEntry->FwVersion = Version->ImageVersion.SecureVerNum;
+    pEsrtEntry->LowestSupportedFwVersion = Version->ImageVersion.SecureVerNum;
+  }
+
+  //
+  // Get status of last firmware update attempt
+  //
+  FwUpdateStatus = (FIRMWARE_UPDATE_STATUS *)RsvdBase;
+
+  if (FwUpdateStatus->Signature != 0xFFFFFFFF) {
```

Comment:
I think it is better to use  if (FwUpdateStatus->Signature == ????????)  instead.